### DISCUSSION TITLE
建议更换仓库中的Socat插件,因为他在23.05版本中无法兼容firewall4

### DISCUSSION THREAD

**Discussion by toss-a** [immortalwrt/immortalwrt#1378] [Ideas]
https://github.com/immortalwrt/luci/tree/openwrt-23.05/applications/luci-app-socat
这个目录下的Socat插件添加端口转发后,发现没有自动打开防火墙端口,看起来似乎他并不兼容firewall4防火墙,使用以下项目的Socat插件得到解决,希望能够更换
https://github.com/chenmozhijin/luci-app-socat

**Comment by 1715173329** [comment]
为什么不直接使用防火墙集成的端口转发功能？

**Reply by toss-a** [reply]
防火墙的端口转发应该不能内网IPv4转发到外网IPv6吧

**Reply by wyndam** [reply]
貌似你这个socat也打不开端口

**Reply by toss-a** [reply]
> 貌似你这个socat也打不开端口

我试过可以,前提是你的ipv6为拨号获取的,防火墙用的firewall4,内网ipv4转发到路由器的拨号ipv6上,你说不行 肯定是你设置的问题,我在immortalwrt和lean的lede上都没问题

**Comment by lynn5713** [comment]
> https://github.com/immortalwrt/luci/tree/openwrt-23.05/applications/luci-app-socat 这个目录下的Socat插件添加端口转发后,发现没有自动打开防火墙端口,看起来似乎他并不兼容firewall4防火墙,使用以下项目的Socat插件得到解决,希望能够更换 https://github.com/chenmozhijin/luci-app-socat

同样的问题，我发现仓库里的socat还是用的iptables，能打开端口，但是不能转发。我的场景跟你是反的，我只有IPV6 公网，想映射到内网的IPV4。
root@ImmortalWrt:/var/etc# uci show firewall | grep socat
firewall.socat=include
firewall.socat.type='script'
firewall.socat.path='/var/etc/socat.include'
firewall.socat.reload='1'
root@ImmortalWrt:/var/etc# uci show socat 
socat.global=global
socat.global.enable='1'
socat.fccafe2f45ab4f9a8e06831397a9c123=config
socat.fccafe2f45ab4f9a8e06831397a9c123.enable='1'
socat.fccafe2f45ab4f9a8e06831397a9c123.firewall_accept='1'
socat.fccafe2f45ab4f9a8e06831397a9c123.remarks='备注'
socat.fccafe2f45ab4f9a8e06831397a9c123.protocol='port_forwards'
socat.fccafe2f45ab4f9a8e06831397a9c123.proto='tcp'
socat.fccafe2f45ab4f9a8e06831397a9c123.listen_port='65533'
socat.fccafe2f45ab4f9a8e06831397a9c123.reuseaddr='1'
socat.fccafe2f45ab4f9a8e06831397a9c123.dest_proto='tcp4'
socat.fccafe2f45ab4f9a8e06831397a9c123.dest_ip='192.168.6.1'
socat.fccafe2f45ab4f9a8e06831397a9c123.dest_port='80'
socat.fccafe2f45ab4f9a8e06831397a9c123.family='6'
root@ImmortalWrt:/var/etc# cat /var/etc/socat.include 
#!/bin/sh
iptables-save -c | grep -v "SOCAT" | iptables-restore -c
iptables-restore -n <<-EOT
*filter
COMMIT
EOT
ip6tables-save -c | grep -v "SOCAT" | ip6tables-restore -c
ip6tables-restore -n <<-EOT
*filter
COMMIT
EOT

**Comment by fairok** [comment]
依然碰到一样的问题，端口并没有打开

**Comment by fairok** [comment]
> https://github.com/immortalwrt/luci/tree/openwrt-23.05/applications/luci-app-socat 这个目录下的Socat插件添加端口转发后,发现没有自动打开防火墙端口,看起来似乎他并不兼容firewall4防火墙,使用以下项目的Socat插件得到解决,希望能够更换 https://github.com/chenmozhijin/luci-app-socat

我们貌似可以在diy_part2.sh里面编写

**Reply by toss-a** [reply]
保存配置后重启immortalwrt或者在启动项重启luci_socat和socat就自动添加了防火墙规则,master分支看起来已合并最新的luci,不必做任何修改

**Reply by fairok** [reply]
> 保存配置后重启immortalwrt或者在启动项重启luci_socat和socat就自动添加了防火墙规则,master分支看起来已合并最新的luci,不必做任何修改

我用昨晚git最新代码编译的固件，重启后在内网ipv6+端口号能访问，在外网ipv6+端口号不能访问，猜测还是端口没打开的问题，请问一下大佬打开端口会不会自动在防火墙里加规则？我看到我的没有自动加规则。

**Reply by toss-a** [reply]
会自动加,如果不会就试试我那个链接上的,记得把feeds里面的luci删掉然后克隆我那链接的,我昨天编译的时候我看有几个没有汉化就换了我发那链接的luci,或者你可以试试手动放规则.
![QQ20241024-181456](https://github.com/user-attachments/assets/1b780863-ffe0-4fd0-bdbc-8ac7a5f3ba6b)

**Comment by baysonfox** [comment]
做个补充：手动把feeds/luci/applications下的 luci-app-socat 替换为https://github.com/chenmozhijin/luci-app-socat 后，公网和内网使用v6+端口都一切正常.

**Reply by itispip** [reply]
> 做个补充：手动把feeds/luci/applications下的 luci-app-socat 替换为https://github.com/chenmozhijin/luci-app-socat 后，公网和内网使用v6+端口都一切正常.

请问怎么替换？尝试把ipk下载下来，通过openwrt的upload package安装，不能运行。

**Reply by baysonfox** [reply]
我是直接从源码编译整个固件的，直接替换了包的源码

**Reply by WROIATE** [reply]
我去年还专门自己根据lienol的改了一版，也可以用，这版是直接用nft+procd的
https://github.com/WROIATE/luci-app-socat

**Reply by Samuel-Chia** [reply]
> 我去年还专门自己根据lienol的改了一版，也可以用，这版是直接用nft+procd的 https://github.com/WROIATE/luci-app-socat

大佬有编译好的软件包么，小白刚从iStoreOS切换到immortalwrt，还不太会编译

**Reply by fairok** [reply]
用lucky也可以

**Comment by 1715173329** [comment]
suspended by https://github.com/immortalwrt/luci/commit/2fcc26e876ce013fda3d1ff46a80ecfaadf2f6ac

**Reply by toss-a** [reply]
试了下可以用 但是勾选防火墙后也没有自动添加规则，在防火墙规则里面也没有看到相关的放行规则，需要在防火墙里面手动添加规则，测试版本23.05

**Reply by Samuel-Chia** [reply]
我自改的socat增加了子进程数量限制和UDP超时时间，为了解决rustdesk的UDP21116端口创建大量子进程的问题。
大佬这个用的3proxy，作为代理软件应该不会有类似问题了。第一次知道还有3proxy这个玩意可以替换socat。不知道支不支持proxy protocol

**Reply by 1715173329** [reply]
> 试了下可以用 但是勾选防火墙后也没有自动添加规则，在防火墙规则里面也没有看到相关的放行规则，需要在防火墙里面手动添加规则，测试版本23.05

你没看到不代表没有。事实上是根本不需要那种做法。

**Reply by 1715173329** [reply]
> 不知道支不支持proxy protocol

看上去是在三月份加了 protocol v1 支持，不过正好错过 release。

**Reply by 1715173329** [reply]
> 试了下可以用 但是勾选防火墙后也没有自动添加规则，在防火墙规则里面也没有看到相关的放行规则，需要在防火墙里面手动添加规则，测试版本23.05

你遇到的问题大概是防火墙没重载导致的，自己加一下即可
https://github.com/immortalwrt/luci/commit/186ecbd53390d0f06f7b74dff52dd480328e48fc

**Reply by toss-a** [reply]
> > 试了下可以用但是勾选防火墙后也没有自动添加规则，在防火墙规则里面也没有看到相关的放行规则，需要在防火墙里面手动添加规则，测试版本23.05
> 
> 你遇到的问题大概是防火墙没有重载导致的，自己加一下就可以 [immortalwrt/luci@ 186ecbd](https://github.com/immortalwrt/luci/commit/186ecbd53390d0f06f7b74dff52dd480328e48fc)

添加后完美解决，谢谢大佬
